target/linux/mpc85xx/image/p1020.mk

```diff
@@ -15,6 +15,17 @@ define Device/aerohive_hiveap-330
 endef
 TARGET_DEVICES += aerohive_hiveap-330
 
+define Device/extreme-networks_ws-ap3825i
+  DEVICE_VENDOR := Extreme Networks
+  DEVICE_MODEL := WS-AP3825i
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
+  BLOCKSIZE := 128k
+  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
```

Comment:
In the commit message above, you mention

> The flattened device tree must be resized multiple times to deal with
  U-boot's pickiness about making space for the fdt without relocating
  it. Although `help fdt` advertises that `fdt resize` pads to a 4K-aligned
  address, it cannot predict the future size of the fdt after later `fdt`
  operations.

This sounds a lot like the troubles I had with apm821xx, uboot couldn't add/extend nodes.

The solution there was to tell the dtc compiler to pad the dtb with empty space.

apm821xx uses a special setup for this. It overwrites the "dtb" with its own version that passes a new DEVICE_VAR DTB_SIZE which does the rest. (the 32KiB is a bit too generous.)

https://github.com/openwrt/openwrt/blob/master/target/linux/apm821xx/image/Makefile#L31

Maybe this works here as well and you can get rid of the ranges that uboot is supposed to fill?

Reply:
Any luck with that method? Now that the AP-350 is "out of the way"?

Reply:
I'm attempting a compile now (this is mostly a comprehension test for me on how image recipe definition lines work; we'll see if I pass).

Reply:
It works! I need to refine things a bit, but here's what I tried so far:
```diff
diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
index 2b81884032..e1d1bc6948 100644
--- a/target/linux/mpc85xx/image/Makefile
+++ b/target/linux/mpc85xx/image/Makefile
@@ -3,6 +3,12 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/image.mk
 
+DEVICE_VARS += DTB_SIZE
+
+define Build/dtb
+       $(call Image/BuildDTB,$(DTS_DIR)/$(DEVICE_DTS).dts,$@.dtb,,--space $(DTB_SIZE))
+endef
+
 define Device/Default
   PROFILES := Default
   DEVICE_DTS := $(lastword $(subst _, ,$(1)))
diff --git a/target/linux/mpc85xx/image/p1020.mk b/target/linux/mpc85xx/image/p1020.mk
index 93fd315e36..feb327036d 100644
--- a/target/linux/mpc85xx/image/p1020.mk
+++ b/target/linux/mpc85xx/image/p1020.mk
@@ -63,7 +63,8 @@ define Device/extreme-networks_ws-ap3825i
   DEVICE_MODEL := WS-AP3825i
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
   BLOCKSIZE := 128k
-  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
+  DTB_SIZE := 20480
+  KERNEL = dtb | kernel-bin | lzma | dtb | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
   IMAGES := sysupgrade.bin
   IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
 endef
```

Reply:
A combination of

```txt
interrupts off;
bootm start 0x2000000;
bootm loados;
fdt resize;
fdt boardsetup;
fdt chosen;
bootm prep;
bootm go;
```

... and
```diff
diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
index 2b81884032..e1d1bc6948 100644
--- a/target/linux/mpc85xx/image/Makefile
+++ b/target/linux/mpc85xx/image/Makefile
@@ -3,6 +3,12 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/image.mk
 
+DEVICE_VARS += DTB_SIZE
+
+define Build/dtb
+       $(call Image/BuildDTB,$(DTS_DIR)/$(DEVICE_DTS).dts,$@.dtb,,--space $(DTB_SIZE))
+endef
+
 define Device/Default
   PROFILES := Default
   DEVICE_DTS := $(lastword $(subst _, ,$(1)))
diff --git a/target/linux/mpc85xx/image/p1020.mk b/target/linux/mpc85xx/image/p1020.mk
index 93fd315e36..5049847cb8 100644
--- a/target/linux/mpc85xx/image/p1020.mk
+++ b/target/linux/mpc85xx/image/p1020.mk
@@ -63,7 +63,8 @@ define Device/extreme-networks_ws-ap3825i
   DEVICE_MODEL := WS-AP3825i
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
   BLOCKSIZE := 128k
-  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
+  DTB_SIZE := 20480
+  KERNEL = kernel-bin | lzma | dtb | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
   IMAGES := sysupgrade.bin
   IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
 endef
```

... is enough to avoid repeated resize. I will adjust my commit to match.

Reply:
Fixed, and pushed in force-push 557192907b9652b21f1cc8890b807edae22aed40.